src/main/java/InfrastructureManager/MatchMaking/ScoreBasedMM.java

```diff
@@ -0,0 +1,228 @@
+ * Weight: 1 - 10 - 10 - 10
+ *
+ * @author Zero
+ */
+public class ScoreBasedMM implements MatchMakingAlgorithm {
+
+
+    private final Logger logger = LoggerFactory.getLogger(this.getClass());
+    // TODO: this should be dynamic
+    private static final long ACCEPTABLE_PING = 300; //use this as parameter before put them in the calculation
+    private static final long RESOURCE_WEIGHT = 10;
+    private static final long NETWORK_WEIGHT = 10;
+    private static final long PING_WEIGHT = 1;
+    private static final long HISTORY_WEIGHT = 10;
+    private static final long QOS_THRESHOLD = 100000;
+
+    Multimap<String, HashMap<String, Long>> nodeHistory;
+    EdgeClientHistory clientHistory;
+    long deductScorePerSecond = 1;
+
+    /* big step
+    DONE: Score = Wd_D + Wc_C + Wn_N + Wh_H --- D = ping, C = res, N = network, H = history
+        Weight: 1 - 10 - 10 - 10
+    DONE: implement slots (resource and network), this should be an attribute in edge client and edge note.
+    TODO: History implementation, using SQL? Or keep throwing JSON object back and forward?
+    DONE: Need to make sure there won't be no duplicate in nodeList (Id and Address)
+    TODO: Weight should be dynamic later on
+    TODO: Implement time
+
+     */
+
+    /*
+    TODO: implement visibility: blind-inacc-accurate. This rather connected to "input" analyze,
+     take the input and somehow "blur" the info by introduce some deviant (multiply by 5-30%, so we get the estimation)
+
+     */
+
+    /* small step
+    DONE: match base on available resources
+    DONE: eliminate nodes that can't accommodate client (not enough resources, network etc)
+    DONE: compare between "acceptable" node using score
+     */
+    @Override
+    public EdgeNode match(EdgeClient thisClient, List<EdgeNode> nodeListInput, EdgeClientHistory thisClientHistory) {
+        logger.info("""
+                
+                ----------------------------MATCH MAKING - SCORE BASED--------------------
+                R_weight, N_weight, D_weight, H_weight, QOS_threshold
+                {}      , {}      , {}      , {}      , {}    """
+                , RESOURCE_WEIGHT, NETWORK_WEIGHT, PING_WEIGHT, HISTORY_WEIGHT, QOS_THRESHOLD);
+
+        //Initiating variable
+        List<EdgeNode> nodeList = new ArrayList<>(nodeListInput); //In case of multi threading
+        EdgeNode bestNode = new EdgeNode();
+        EdgeNode rejectNode = new EdgeNode("rejectNode", "000.000.000.000", false, Long.MAX_VALUE, Long.MAX_VALUE, Long.MAX_VALUE, Long.MAX_VALUE, Long.MAX_VALUE); //return this if score not good
+        long bestScore = 0;
+
+        this.clientHistory = thisClientHistory;
+
+        //initiate temp/comparing variable
+
+        int numberOfUnqualified = 0; //to count the number of node we have been ruled out during iteration
+        int totalNumberOfNode = nodeList.size();
+
+        //Get client requirement (required resource, network)
+        String thisClientID = thisClient.getId();
```

Comment:
So if we have 100 pairs. Every time we update a node (which connected to client(s)), not only we have to update it in the nodeList, we also have to access the class `ScoreEntry` and replace the values in the matrix (not sure how to access them easily tbh)
Don't you think that will create more trouble? Compare to just getting a variable in an object?

Reply:
And that is why references are used, you shouldn't copy the node data into the new class, just add the node as a variable itself, and any changes to the node will be reflected in the score as well.

Reply:
I got what you mean now, rather than create a new variable with the same name, using the address would be better.
I will do this after we kind settled down with the abstraction/definition of match making algorithm.
Going to leave a TODO here for now
Done in a283307